Default iface per bridge

overhacked opened the pull request:
Adds ability to specify a `vnet_default_iface` per bridge in the `interfaces` property. Helps make configuration more expressive for the proposed netgraph PR #843.

Instead of:
```
interfaces=vnet0:bridge0,vnet1:bridge1
```
and adding host interfaces in `rc.conf`, one can specify, for example:
```
interfaces=vnet0:bridge0:em0,vnet1:bridge1:tap1
```

Make sure to follow and check these boxes before submitting a PR! Thank you.

- [x] Explain the feature
- [x] Read [CONTRIBUTING.md](https://github.com/iocage/iocage/blob/master/CONTRIBUTING.md)

sonicaj committed:
Improve validation for VNET MAC Fields

This commit improves validation for vnet mac props by ensuring that a valid mac address is specified each time with taking into account that the addresses aren't repeated.
Ticket: #45838

sonicaj committed:
Make default interface for vnet customizable

This commit adds support for customizing the default interface for vnet which should be attached on the relevant bridge. It updates the properties of the jails by adding a new property which is used when starting vnet jails.
Ticket: #40484

sonicaj committed:
This commit adds support for a default none value for vnet_default_interface property

sonicaj committed:
Bug fix for listing jails

This commit fixes a bug where in case a jail's networking was broken, we wouldn't be able to list down all the jails as that jail would leave a traceback.
Ticket: #47151
Closes #599

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
Merge branch 'waqar_commits'

Commit:
Add failed_dhcp for this error

Commit:
Update py-libzfs

Commit:
3.6 -> 3

FreeNAS has 3.7, hilarity ensued.

Commit:
And these!

Commit:
3.6 -> 3 (#641)

* 3.6 -> 3

FreeNAS has 3.7, hilarity ensued.

Ticket: #50123

* And these!

Commit:
Large rework to Exec interface

- Use python dns.resolver
- Drop other types of DNS tests
- Make Exec a generator
- New Exception type CommandFailed for Exec failures
- Use Exec for RELEASE updating along with jail
- Break console out of Exec
- Log update progress now (means colors as well)

FreeNAS Ticket: #37742

jsegaert committed:
Added package dnspython (#645)

thedix committed:
Fixed building py-libzfs after updating (#651)

jsegaert committed:
iocage console hangs (#653)

Commit:
Avoid false positive with EOL notice on jail updates

This avoids the exception that would prevent successful jail updates.
- Change python binary to be less static for easier testing on FreeNAS

Commit:
Add back the DNS checks from the jail side for plugins

FreeNAS Ticket: #37742

Commit:
Simplify Exec rework, make CommandFailed return an iterable

This commit also has the exec cli now correctly returning exit codes and will not block on empty stdout/stderr

Commit:
Make Exec a generator

Another massive change to exec, simplifying the interface even more.

- Add interactive arg to iocage API (for exec cli)
- consume_and_log method added to help sanitize text easier for printing/logging
- Change to all IOCExec usages to use the generators
- Show the full domain instead of a quick name for DNS checks
- Fix previous behavior hanging on empty or blank stdout/stderr
- Fix bad interactive commands like top with exec

FreeNAS TIcket: #37742

jsegaert committed:
Add support for placeholders in adminportal (#659)

ole-db committed:
if 'vnet_default_interface' is set to 'none' no host interface will be added to the bridge (#655)

* if 'vnet_default_interface' is set to 'none' no host interface will be added to the bridge. Auto-detected interface is added to bridge if 'vnet_default_interface' is set to 'auto'

* if 'vnet_default_interface' is set to 'none' don't create empty bridge'

Commit:
Don't raise a pkg error if it succeeds after the first install

FreeNAS Ticket: #51654

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
We now call the bridge_cmd if the variable is added

Otherwise we run into missing bridge_cmd's variable.

Commit:
More Exec work

- Simplify consume_and_log to always return a list with stdout and allow disabling logging.
- Simplify generator
- Add a SilentExec class to simplify cases where we don't want any output, not even a list.

FreeNAS Ticket: #37742

Commit:
Do not silence plugin property setting

Spotted by @jsegaert

Commit:
Improve DNS exception for Plugins

Commit:
Have Upgrade use Exec interface

- Change pkg variable to unjailed to represent it's actual use
- Allow stdin to be used with Exec
- Close stdout and stderr in exec, stdin will be closed if specified
- Update comment to further clarify intent

FreeNAS Ticket: #50167

Commit:
Use Exec interface for basejail upgrades

Commit:
Ignore callback for exceptions

iocage relies on these being handled, and the other callbacks might not care.

FreeNAS Ticket: #52459

Commit:
Catch any exceptions trying to load freebsd-version during check_config

FreeNAS Ticket: #52637

sonicaj committed:
Delete jails for all exceptions in case of jail failure by default (#664)

This commit makes sure that jails are deleted by default in all exceptions that occur.
Ticket: #50519

Commit:
Skip RELEASE checking for EMPTY jails

During configuration upgrade, this will cause real headaches.

FreeNAS Ticket: #52637

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
Simplify config upgrades

It was a little too verbose, this is nicer to read.

ole-db committed:
adds 'thickjail' to jailtypes.rst (#666)

* adds 'thickjail' to jailtypes.rst

* typo

a-stjohn committed:
Made changes to iocage docs. Came from ticket48768. (#663)

* Made changes to iocage docs. Came from ticket48768.

* Updated with regard to feedback.

* Updated with regard to feedback.

Commit:
Add TERM to default clean envs

In addition we also no longer set None for a clean env on exec/console. Not setting TERM resulted in some odd entries like su or screen that don't always play nice with utilities inside the jail.

FreeNAS Ticket: #52237

jsegaert committed:
Quietly start jail if needed when running 'iocage exec' (#672)

ole-db committed:
fixes confusion with on/off and yes/no (#673)

sonicaj committed:
Raise proper exception for old invalid Mac addresses (#667)

This commit makes sure that we raise exception via proper iocage channels incase an old invalid mac address entry exists in the system.
Ticket: #53674

Commit:
Add dnspython to reqs

thedix committed:
Fixed sorting jails at iocage get -r (#675)

Commit:
Rewrite how devfs_rulesets are generated

- Simplify plugin devfs ruleset generation
- Rewrite generate_devfs_ruleset

We now no longer touch the hosts /etc/devfs.rules, instead we create a dynamic ruleset that we clean up on jail destruction.

This ends up fixing some bugs where setting dhcp after setting allow_tun or vice versa would not work.

In addition, tuns can now properly be created in the target jail, as before the devfs rules were not being correctly reloaded.

FreeNAS Ticket: #40872

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
There's nothing to see here!

Seriously, why'd you look?

Commit:
Always generate a devfs ruleset for a jail

- We no longer rely on the hosts devfs_ruleset of 4 to be valid. We hardcode and apply the default 4 rules of a clean host to ensure jail safety
-If a user supplies a devfs_entry, this hardcoding will not occur, as long as that entry exists
- Simplify devfs_ruleset generation during start
- Make sure we are actually removing our devfs rulesets on stop, and not one the user created with the intention of using them with multiple jails. Including the hosts defined 0-4 rulesets

FreeNAS Ticket: #40872

Commit:
Rewrite this more succinctly courtesy of @danjacques

Also remove an unused variable.

Commit:
Handle corrupt default configurations

The user is now told when the default configuration is created or if they have a corrupted one.  In which case iocage will use the normal defaults in memory instead of committing to a file so the user can keep their settings but fix the corruption.

FreeNAS Ticket: #54861

Commit:
Bump to RC

Commit:
Debugs now correctly start fresh

Previous behavior infinitely appended data.

Also add netstat to host debug and correct df header

FreeNAS Ticket: #56202

Commit:
Continue on corrupt jail configurations

Previous behavior was to immediately die

- Add 3 new exceptions for misconfigured jails
- Start to use them in ioc_json
- Show corrupt jails in iocage list
- Return corrupt jails in get all recursive API with default key values being replaced with N/A (FreeNAS focused)

FreeNAS Ticket: #51997

Commit:
Include templates in debug

FreeNAS Ticket: #56232

Commit:
Properly set rtsold during creation

FreeNAS Ticket: #54825

Commit:
spotted by @jsegaert

*cough*

Commit:
Warn about dhcp/accept_rtadv during create if required props are missing

Also fix unbound key spotted by @thedix

Commit:
Fix debug for corrupt jails

We also shift from overloading the UUID to setting status to CORRUPT. Except for one specific instance of list -q, as that doesn't display the status column.

FreeNAS Ticket: #51997

Commit:
Add accept_rtadv and autoconf to jail nic when accept_rtadv is set

FreeNAS Ticket: #54825

jsegaert committed:
Fix quirks with 'iocage help' (#679)

asomers committed:
Optionally allow mounting fusefs from within the jail. (#681)

* Optionally allow mounting fusefs from within the jail.

* Disable allow.mount.fusefs on older FreeBSD versions

jsegaert committed:
Allow dots in the name for 'iocage console' (issue 677) (#683)

thedix committed:
Fixed bug: do not umount path if fstab entry is not found (#684)

Commit:
Fix bad indent

This didn't allow dhcp enabled jails to actually start dhclient.

FreeNAS Ticket: #56844

a-stjohn committed:
Fixed error message for corrupted default.json (ticket 56364). (#685)

utrenkner committed:
Drop "zero dependency" from description (#686)

* Drop "zero dependency" from description

This still stemmed from the old shell based iocage.

* Remove "drop in" from description

Not quite correct anymore since moving to Python...

Commit:
Use default gateway for mac_prefix or generate one


FreeNAS Ticket: #56556

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
Add missing return for setting default prop

This avoids a harmless jail not found error that was erroneously throwing on default configuration setting.

Commit:
Make import more robust with handling naming

Jails created with a _ in their name (plugins for example) would fail as the split was too greedy.

FreeNAS Ticket: #56211

Commit:
Add enumeration to internal sorting for RELEASEs

This would have user RELEASE datasets that reused the major of a release to either override or be overridden by the RELEASE.

FreeNAS Ticket: #57861
Closes #687

Commit:
Set templates to read/write for json writing and back to readonly

This alleviates issues of configuration version incrementing along with possibly wanting to remove properties that were set. Like ip4_addr and such.

FreeNAS TIcket: #57876
Closes #625

Commit:
Write partial configurations by default

- Configurations now show what the user has set, with a few jail specific properties
- Removed many instances of passing configuration to certain classes
- We now merge the default props with the jail props when reading them in via json_get_value('all')

Commit:
Fallback to using the jails freebsd_version if the RELEASE one fails

Thick/Empty jails and warden migration jails from FreeNAS fall in this category.

FreeNAS Ticket: #58365

Commit:
Remove unnecessary notes variable

Commit:
Skip checking configuration if the versions match

This avoids wasted time going through the motions. Also allow users to use list, get and such without requiring root unless a configuration version changed (spotted by @jsegaert)

FreeNAS Ticket: #58617

jsegaert committed:
Need 'type' for plugins

Commit:
Correctly add bpf to devfs rules, change creation warnings for props

Instead of warning the user they need these, we set them for the user in addition to the warning.

Also show all missing props during start.

FreeNAS Ticket: #53334

Commit:
Correct typo

Commit:
Fix lack of writing json if json_load returned True

Also corrects incorrect parameter assignments.

Commit:
Missed an instance

Commit:
Fix tests and default configuration loading in ioc_json

- Previous behavior in various locations would open the defaults file directly, it should be using the checking mechanism.
- Add a unit_test attribute to IOCStart for the unit test we have currently
- Fix running tests since the module flattening
- Update RELEASE to 11.2-RELEASE by default for fetch test
- Use only base.txz for fetch test to speed up efforts as the others are unused

FreeNAS Ticket: #59004

Commit:
Use en_US.UTF-8 or users LANG by default for console/exec

Closes #669
FreeNAS Ticket: #59193

Commit:
Remove backup tests

These aren't needed

Commit:
Fix bad path being sent to upgrade

FreeNAS Ticket #59751

Commit:
Allow listing only basejails with -B

Commit:
Add basejail header to long list

Shows if the jail is a basejail or not.

Commit:
Use the first item for dicts sent to get recursive

FreeNAS Ticket: #59823

Commit:
Show devfs ruleset when starting and return correct one

If the jail is running, the devfs ruleset will never be correct unless a user is supplying a manual one.

Closes #694
FreeNAS Ticket: #59841

Commit:
FIx get -r

Commit:
Update now correctly updates the release property with the patch level

In addition basejails will now all see the new patch level upon one updating that shares the same release.

FreeNAS Ticket: #59442

Commit:
Correctly mount basejail and umount basejails during cloning

FreeNAS Ticket: #60165

Commit:
Move updating basejails to base module ioc_fetch

This has the additional advantage of updating basejails when the release that is fetched is updated.

Improvement suggested by @jsegaert.

FreeNAS Ticket: #59442

Commit:
Fix hostname in rc.conf and /etc/hosts generation

We were missing a split and the rc.conf entry had a typo.

FreeNAS Ticket: #58794

Commit:
Fix pkglist and no IP options fallout with thin configuration

FreeNAS Ticket: #60507

Commit:
Destroy leftover snapshot for thick jail creation

FreeNAS Ticket: #60525

Commit:
Unneeded cloned_release property kept around with thick jails

Remove unused variable in IOCUpgrade

FreeNAS Ticket: #60534

Commit:
Move progress out of update_progress for fetch

Also send callbacks if one was supplied even if silent is set.

FreeNAS Ticket: #54717

thedix committed:
Fixed bug: get non-existing keys from conf (#701)

Commit:
Break ioc_json into more classes

Possibly more can be moved into the new classes, but for now these were needed.

- Fix bad release prop setting
- Reduce all usage of getting pool and iocroot many times

Closes #700
FreeNAS Ticket: #60768

Commit:
Remove unneeded comment

Commit:
Add thick configuration for creation of jails and plugins

This will cause the configuration to not inherit the hosts defaults.json, which is the default behavior.

FreeNAS Ticket: #60774

Commit:
Remove unused self.zfs instance per @sonicaj review

Commit:
Address more @sonicaj review

Commit:
Errrrr

Commit:
Remove identical if statement

Commit:
Address even more @sonicaj concerns

Reviews are good!

sonicaj committed:
Bug fix for iocage fetch (#706)

This commit fixes a bug which was affecting fetch from fetching releases and also thick config plugins couldn't be created.
Ticket: #61191

sonicaj committed:
Bug fix for always creating jails with name and count flags (#708)

This commit fixes a bug where we were not able to create count jails when a valid uuid was supplied as a name. Apart from that this commit enforces that count flag cannot be used with uuid flag and removes force option from create command as it is not used anywhere.
Ticket: #61551

Commit:
Remove unused line

Commit:
Join multiline error messages with logit

Commit:
Create new exception class for inheritance.

Also create

Commit:
Massive refactor of ioc_fstab

- Create reusable single instance of fstab
- Validates fstab before loading it now (unless using list, that's for users to spot the issues themselves)
- Check for duplicate entries
- Check if destination exists
- Check if source/destination is a directory
- Check if source/destination is an absolute path
- Check if dump/pass are single proper integers

Closes #702

Commit:
Bump py-libzfs to latest

Commit:
Properly install requirements.txt

Closes #652

Commit:
Remove unneeded code

Commit:
Allow auto to be set for vnet_default_interface

FreeNAS Ticket: #61974

sonicaj committed:
Raise DHCP error if DHCP not configured (#709)

This commit fixes a bug where if 0.0.0.0 was the ip address acquired by the jail, an error was not raised by iocage.
Ticket: #61983

Commit:
Reads nicer!

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
Check for malformed fstab lines

Commit:
Use default rc.conf file instead of builtin boilerplate

This allows the user to customize their rc.conf that is deployed to jails. Important to customize for environments.

Behavior:
- If jail is being created from a template, it will use the templates rc.conf instead of the default_rc.conf. 
- If jail is being created from a RELEASE, it will use the default_rc.conf located in the iocage mountpoint.

Closes #615

Commit:
Fix bad conditional on default rc.conf usage

Commit:
Default to auto for vnet_default_interface

Most users will want this behavior, with thin jails, it's an easy change to go and use none for most.

Ticket: #62361

Commit:
Catch users using old vnet_default_interface prop value of none

This was auto, it's now the case :)

FreeNAS Ticket: #62361

Commit:
Catch version 12 too

Commit:
Re-set self.rc on restart

FreeNAS Ticket: #61659

Commit:
Fix self.default_config not always existing

Old debug

Commit:
Add missing pathlib.Path for freebsd_version

sonicaj committed:
Destroy Network related setup on stop (#713)

This commit changes previous behaviour for iocage where now it tries to destroy network related setup if a jail is stopped with force flag set to true. However, in that case we suppress exceptions if any and move on.
Ticket: #61668

sonicaj committed:
Bug fix for template jails (#714)

This commit fixes a bug which prevented creating a template jail at the same time the jail was being created for the first time. The dataset was readonly at that point and system tried to complete the relevant configuration files but failed to do so because of the readonly property set to true.
Ticket: #62190

Commit:
Nicer!

@william-gr keeps reminding me how ugly my shell^W python is.

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
Do not create default configuration if checking datasets

FreeNAS Ticket: #63099

Commit:
Add templates to quick listing

FreeNAS Ticket: #63432

Commit:
Remove unneeded conditional

Spotted by @sonicaj

jsegaert committed:
Fix 'iocage fetch' sorting (#716)

Commit:
Don't add all parameters to start

This causes issues now in 12.0-RELEASE, and also allows us to not load kernel modules that may or may not exist unless the user specifies them for the jail.

FreeNAS Ticket: #63864

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
Use periods instead of colons for nic name

Closes #717

Commit:
Remove doc.txz from default files

FreeBSD no longer supplies non-empty doc.txz's. If a user wants them, they will need to supply the -F flag and add it.

#472

Commit:
Fix a few issues in extraction/downloading code

- Inform user we are downloading the MANIFEST if missing and using the default FreeBSD server
- No longer try to extract/verify files that are missing from the MANIFEST. This avoids the terrible behavior of then trying to redownload that file and reverify, which would only fail.
- No longer extract all files when a redownload of a dist has happened. This was exacerbated by the above issue.

FreeNAS Ticket: #63891
Closes #472

Commit:
Don't extract if the user doesn't have a MANIFEST and uses custom server

Give the user a sane example that will work, not taking into account if they supplied other flags like --noupdate and such.

Commit:
Remove unused missing for fetch

Commit:
Fix a few more remaining fetch bugs

A couple other long standing issues finally squashed:
1 - When a file was missing and extracted, no others would continue extracting
2 - Missing routine only ran once, now runs 3 times
3 - MANIFESTs that were downloaded weren't read back in  for the hashes, resulting in a second required run.

Commit:
Remove extra space in downloading fetch message

Commit:
Cleanup some fetch code, improve error message

Suggestions by @william-gr

Commit:
Check for exec=on when issuing update/upgrade (#723)

This affects fetched RELEASEs also trying to update

Closes #631
FreeNAS Ticket: #64140

Commit:
tkt-64143: Only capitalize fetch RELEASE if everything is default (#724)

* Only capitalize fetch RELEASE if everything is default

Not a local file directory or custom mirror.

Closes: #626
FreeNAS Ticket: #64143

* Address feedback

Remove the upper, users will need to supply a proper name now. Also make checking if the release is newer best effort, as if the users choice of naming causes issues, it's hard to be sure.

Commit:
Grab doc.txz by default for RELEASEs under 12.0-R (#720)

* Grab doc.txz by default for RELEASEs under 12.0-R

FreeNAS Ticket: #64104

* Handle non-standard naming schemes

* Address feedback

Commit:
Fix creating additional epairs (#721)

- vnet0 was the only interface that renamed them to epairs, now all vnetN variants will (vnet0|ADDR,vnet1|ADDR, etc
- Add which nic inside the jail the vnet device on the host is associated with
- Only add default route to the first nic.

Closes #682
FreeNAS Ticket: #64113

Commit:
Fix listing non-standard RELEASE naming schemes (#725)

tkt-64152: Fix listing non-standard RELEASE naming schemes

Commit:
Properly set mtu on all vnet interfaces (#722)

tkt-64116: Properly set mtu on all vnet interfaces

Commit:
Fix mismatching RELEASEs being fetched, and a couple other bugs (#729)

Noted by @jsegaert

- Use the right interface for fetching instead of internal plumbing
- Fix bad variable usage
- Fix a couple other exceptions that could occur with the fetch interface

FreeNAS Ticket: #64638

Commit:
Supply a files list to IOCFetch with create (#732)

Otherwise the RELEASE will use the default files, and they will not be equivalent to what the user would get fetching from the cli.

FreeNAS Ticket: #64905

Commit:
tkt-64899: Rewrite IOCZFS to use subprocess calls (#731)

* Rewrite IOCZFS to use subprocess calls

libzfs writes to stderr, and messing with fd's is not very clean. This also seems to perform better

FreeNAS Ticket: #64899

* Address review by @william-gr

- Remove zpool property, iocage doesn't use this
- Set the right prop as p

Commit:
Fix travis and all flake8 errors (#733)

Travis was previously broken by c15be3d74c2e9618fa5708d7a2924a096fa11402, this restores the correct behavior and fixes all issues that came since.

FreeNAS Ticket: #64926

sonicaj committed:
tkt-63801: Check snapshot exists before destroying (#735)

* Check snapshot exists before destroying

This commit ensures that we check a snapshot exists before we destroy it and log an appropriate message accordingly. It also introduces a snapshot class which we can use to better manage snapshot related operations.
Ticket: #63801

* Address William's review

* Address reviews

Commit:
tkt-62547: No longer start network services explicitly (#734)

* No longer start network services explicitly

- Use explicit fstab for sysrc during creation
- New syntax for ip4_addr
EXAMPLE : ip4_addr='vnet0|DHCP' 
This replaces dhcp=on, as that previous property would start DHCP on all interfaces, which isn't always desired.
- Simplify setting rc.conf in start to use sysrc for DHCP or rtsold.
- New property 'rtsold', this will enable the service in the jail. This requires having accept_rtadv in the ip6_addr, this new property allows one to rely on just rtsol instead of always starting rtsold.

Closes #711
FreeNAS Ticket: #62547

* Address reviews by @sonicaj and @william-gr

* Yay!

sonicaj committed:
Initial commit fir cirrus-ci

This commit adds support for running the tests in cirrus-ci.

Commit:
Return a more verbose error message on plugin failure (#739)

Also return the last 10 lines instead of just the last one.

FreeNAS Ticket: #65007

Commit:
tkt-65421: Fix snapremove (#745)

* Fix snapremove

After the recent merger and usage of classes a variety of issues popped up with this one.
- Fold snapremove into the API (Avoids greedy matching on names as well)
- Remove defunct class
- Silence stderr of IOCSnapshot
- Clarify comment (updating every func call would defeat the point of instantiating the class imo)
- Correctly return True/False for exists
- Inherit from IOCZFS

FreeNAS Ticket: #65421

* Let's only run this once

Previously we were running this three times for each snapshot.

* Add TODO

sonicaj committed:
Add More tests  (#740)

* Initial commit for adding more tests (#736)

This commit adds the basics for the functional tests coming in.
Ticket: #59247

* Starting/Stopping Jails Tests

This commit adds tests relating to starting and stopping jails

* Setting/Getting Props Tests

This commit adds tests relating to setting and getting props on jails

* Exec/Clone/Debug Tests

This commit adds tests relating to exec, clone and debug commands

* Df/Export/Import Tests

This commit adds tests relating to df, export and import commands

* Console/Rename/Restart Tests

This commit adds tests relating to console, rename and restart commands

* Fstab/Snapshot Tests

This commit adds tests relating to fstab and snapshot related commands

* List Tests

This commit adds tests relating to list command

* Tests for activate and fetch

This commit refines the existing tests for activate and fetch

* Creating Jails tests

This commit adds tests for creating jails

* Clean/Destroy Tests

This commit adds tests relating to destroy and clean commands

* Not stopping at first failed test

This commit fixes a bug where the tests stopped at the first failed test.

sonicaj committed:
Remove py-libzfs as submodule (#747)

This commit removes py-libzfs as a submodule from iocage and instead uses pkg to install py-libzfs.
Ticket: #64332

Commit:
tkt-65454: Fix setting thickjails as a template or cloning them (#746)

* Fix setting thickjails as a template or cloning them

FreeNAS Ticket: #65454

* Allow clones/templates to be created as thickjails

- Cut down long path usage in calls
- Make creating a thickjail a method

FreeNAS Ticket: #65454

* Travis

* Fix typo and cleanup exception

Commit:
Bump remaining copyrights and start preparing for tagged release

Commit:
Replace hostnames with "_" to valid hostnames (#749)

Plugins that have multiple installs will run into this problem.

FreeNAS Ticket: #65610

Commit:
Add back support for 'ALL' to update/upgrade (#748)

FreeNAS Ticket: #65601

jsegaert committed:
Make check for 'latest' repor more robust (#753)

sonicaj committed:
Update Makefile (#751)

* Update Makefile

This commit adds a better error handling for a case where python path supplied is non-existent.

* Update Makefile

Co-Authored-By: sonicaj <waqarahmedjoyia@live.com>

sonicaj committed:
tkt-65628: Bug fix for iocage fetch (#750)

* Bug fix for iocage fetch

This commit fixes a bug where we didn't fetch the host release when enter was pressed.
Ticket: #65628

* Update iocage_lib/ioc_fetch.py

Co-Authored-By: sonicaj <waqarahmedjoyia@live.com>

reezer committed:
Make links their own section (#754)

They are hard to find in the warnings.

Also fix Webside -> Website and make documentation not all-caps.

jsegaert committed:
update/upgrade ALL processes last jail twice (#755)

Commit:
Change travis over to new flatten modules for before

sonicaj committed:
Bug fix for Upgrade (#762)

This commit fixes a bug where upgrades failed.
Ticket: #66087

Commit:
Bump to 1.1 PRERELEASE

phyber committed:
Leverage jail(8) for shared IP networking behaviour (#770)

* ioc_json.py: Add legacy_networking_behaviour config key

This key controls iocage networking behaviour. If it's turned on, iocage
will try to manage:
  - Which interface to add addresses to if an interface isn't specified
    in ip[4|6].addr
  - Tear down of ip4_addr interfaces when stopping a jail

If it's turned off, iocage will not try to manage these things and will
trust FreeBSD and the system administrator to do the right thing.

* ioc_start.py: Hide working out the gw for ip4_addr behind legacy flag

* ioc_stop.py: Hide teardown of aliases behind legacy networking flag

* .gitignore: Add vim *~ files

* ioc_stop.py: Fix syntax error in if

* ioc_json.py: Fix typo in legacy_networking_behaviour check_config

* ioc_json.py: Add allowed values for legacy_networking_behaviour

* ioc_json.py: Don't fixup ip4_addr if legacy behaviour disabled

* ioc_start.py: Change how ip4.addr list is generated

* ioc_start.py: Change how start_paramaters are formed

* ioc_start.py: Write out jail config to conf file after start

* ioc_stop.py: Attempt to use jail config file if it exists

* ioc_start.py: Move jail config writing to own function

jails are now started using the written config

* ioc_start.py: End jail.conf with a new line

* ioc_stop.py: Comment on jail stop and tidy

* ioc_{json,start,stop}.py: remove iocage managed shared IP

This now lets jail(8) manage it fully.

* ioc_stop.py: Squash jail stdout when stopping jails

* ioc_{start,stop}.py: Minor formatting

* ioc_{start,stop}.py: Fix some flake8 issues

* ioc_start.py: Adjusting line indent blind

local flake8 doesn't pick this up somehow.

* ioc_start.py: Simplify ip4.addr generation

* ioc_stop.py: On successful stop, unlink jail config

* ioc_destroy.py: Attempt unlink of jail conf on destroy

* ioc_{destroy,stop}.py: Don't let jail conf unlinking halt things

Commit:
Catch a missing location, bump netifaces dep

jamesog committed:
Add allow.vmm jail parameter (#767)

FreeBSD 12.0-RELEASE has a new jail(8) parameter "allow.vmm" to allow
jails to access vmm(4).

This adds a new allow_vmm config value to iocage to support this and
bumps the config version to 17.

Commit:
If a plugin's MANIFEST cannot be found directly, we glob for it (#773)

FreeNAS Ticket: #65796

Commit:
Bump copyrights to 2019

jsegaert committed:
Fix various Tracebacks for non-root users (#774)

sonicaj committed:
Check for Faulted Pools (#778)

This commit adds support for raising an exception when a FAULTED pool is presented to iocage to activate. As this pool cannot be used, we raise an exception accordingly.
Ticket: #67620

sonicaj committed:
Check Release exists (#777)

* Check Release exists

This commit makes sure we check the release exists before destroying it. If it doesn't, we raise an appropriate exception outlining that it does not exist. It also introduces a new class Resource which we can use to manipulate resources in iocage. However much thought is still needed on how we should proceed with it in the future.
Ticket: #63810

* Address Reviews

sonicaj committed:
Add ignore exceptions flag (#776)

* Add ignore exceptions flag

This commit introduces a flag for suppressing exceptions when starting/stopping jails. This can be useful when we have jails which cannot really start and we don't want the system to exit in this case.
Ticket: #65391

* Address Reviews

sonicaj committed:
Improve exception handling for listing (#775)

This commit improves exception handling for a case when we tried to determine a jail's ip address and failed to parse the output.
Ticket: #65931

Commit:
Silence jail during destroy/clean (#780)

FreeNAS Ticket: #67791

Commit:
tkt-67789: Add default interface to ip(4|6)_addr if alias doesn't exist (#781)

* Add default interface to ip(4|6)_addr if alias doesn't exist

This allows iocage to set up the address as requested without stomping on potential addresses already requested in rc.conf.

FreeNAS Ticket: #67789

* Refactor this

Simplify the loops to just addresses

* Add pflog to list of skipped interfaces

Commit:
Removing an fstab entry wrote an incorrect file (#788)

The file would be lacking newlines for each entry, which broke fstab -l

FreeNAS Ticket: #68520
Closes #779

cernst1980 committed:
BUG: SYSV settings are ignored (#785)

* BUG: SYSV settings are ignored

When setting sysvmsg, sysvsem and sysvshm to value of "new", jail starts without these settings.
The settings were skipped because of "if '1' in x".

* Update iocage_lib/ioc_start.py

Co-Authored-By: cernst1980 <cernst1980@gmail.com>

* Corrected requested change

sonicaj committed:
tkt-65553: Add More Tests (#759)

* Bug fixes for snapshot related tests

This commit fixes a bug where snapshot tests for template and cloned jails failed sometimes because they have other snapshots associated with them as well which we were not taking into account before

* This commit introduces the following changes:

1) Add tests to verify dhcp jails creation and assign ip's manually
2) Fix issues with snapshot related tests where tests took into account the cloned jail snapshots
3) Minor other fixes

* Indent fixes

* Add tests for upgrade

* Reset execprestop value after testing it as other failing tests might use it

* Address Reviews(1)

* Fixes

* Add tests for update

* Address Reviews

* Minor changes

* Parse multiple ip addresses provided correctly

Commit:
Throw unjailed exceptions with IOCExec (#789)

FreeNAS Ticket: #68961

Commit:
Update SilentExec with changes (#790)

FreeNAS Ticket: #68961

sonicaj committed:
tkt-67800: Jail Configuration Class (#782)

* Introduce class for managing jail configuration files

This commit introduces a class for better managing jail configuration files in iocage.

* Bug fix for exec_prestop

This commit introduces the following changes:
1) Make sure we don't perform post stop operations if we failed to kill the jail process.
2) Bug fix for a case where we performed some exec* operations twice.

* Minor edits

* Address Reviews

* Refactor JailConfiguration class name

* Close jail if exec_poststart fails

* Minor improvements

* travis fixes

* Address Reviews(2)

* Refactor safe_checkoutput method to use IOCExec

* Address Reviews

* Extend SilentExec to return stdout and stderr

Commit:
Correct net property ip4/ip6 to valid jail(8) opts (#791)

sonicaj committed:
Add Validation for ip addresses (#792)

Add Validation for ip addresses

This commit adds validation for ip addresses.
Ticket: #68572

Commit:
tkt-69213: Honor jail user with exec (#796)

* Honor jail user with exec

FreeNAS Ticket: #69213
Closes #795

* Add tests and update README

a-stjohn committed:
Updated man page with hostid_strict_check option. (#798)

sonicaj committed:
Bug fix for jail ips (#804)

Bug fix for jail ips

This commit fixes a bug where we did not configure a supplied ip correctly ( if it wasn't already aliased ) and it was in the format 'interface|ip,ip'.
Ticket: #69724

Commit:
The dragons have gone back into their caves

Occasionally they pop out still, but they are better managed now.

sonicaj committed:
Fix for sorting with ip (#803)

Fix for sorting with ip

This commit fixes an issue where we were not able to correctly sort ips when no interface was supplied.
Ticket: #69717

sonicaj committed:
Bug fix for ip aliases

This commit handles KeyError's raised by various interfaces if a configured mode is not present for an interface.
Ticket: #69996

Closes #807

sonicaj committed:
Improve exception handling (#802)

Improve exception handling

This commit improves how we handle exception handling and also fixes an issue where we could not return values from existing exception handler.

sonicaj committed:
Merge pull request #808 from iocage/issues/69996

tkt-69996: Bug fix for ip aliases

Commit:
Require a mountpoint on parent iocage dataset (#811)

FreeNAS Ticket: #70122
Closes #810

Commit:
tkt-69267: Add --debug(-d) and --images(-i) flag to clean (#797)

* Add --debug(-d) and --images(-i) flag to clean

Also switch over to subprocess zfs as occasionally the base dataset would not be destroyed properly by py-libzfs.

FreeNAS Ticket: #69267

* Add new property to template created jails

- This allows us to destroy jails faster that are from templates.
- Destroy leftovers properly now
- Add iocroot_datasets property
- Remove context manager for zfs_get_property as it's not functioning, Waqar will have a PR for this and the other usages.

* Clarify message on debug cleaning

* Fix clean breaking from destroy behavior change

* Remove debug

* Fix thickjails key

* Use stdout, stderr instead of capture_output for python3.6

* Handle jails named 'root' and fix a string typo

* More fixes

- Use ioc_exceptions to suppress
- Properly handle jails named 'root' for templates
- Destroy leftover parent datasets of jails when a template is destroyed
- Reverse dataset direction to destroy in the correct order
- Add mountpoint special handling to zfs_get_property
- Use zfs_get_property for the hack88 portion to avoid superfluous messages about non-existent datasets during cleaning.

* Travis

* Add callback to IOCZFS

* Travis

sonicaj committed:
Improve Makefile

This commit introduces the following changes:
1) Installs python36 if python provided path does not exist
2) Removes a compulsion for pkg to be using latest instead of quarterly with a warning
3) Make sure we don't install any python2 deps

FreeNAS Ticket: #71103
Closes #813
Closes #814
Closes #815

sonicaj committed:
Merge pull request #818 from iocage/issues/makefile

Improve Makefile

Commit:
Change this to fit the rest (#820)

Commit:
tkt-70977: Add interactive upgrade variable (#817)

* Add interactive upgrade variable

This will sidestep a lot of issues that came up from the one stdin mode we had for noninteractive cases.

- Add InteractiveExec class
- Workaround for master issues with fetching freebsd-update
- No longer fetch master freebsd-update scripts for upgrade, but instead grab the target RELEASEs version of the script
- Add interactive switch and change behavior depending on mode
- Remove a lot of unneeded code now from interactive commands (from iocage API side) and console.

FreeNAS Ticket: #70977
Closes #771
Closes #769

* Fix handful of issues with exec

- {jail|host}_user not being respected
- Skip the starting messages when using interactiveexec in API
- Fix tests to reflect that jails will be started for exec

* Travis

* Fix clone jail tests

Commit:
Fix erroneous template messages on creation failure (#822)

FreeNAS Ticket: #71587
Closes #809

Commit:
Allow non-standard -RELEASE naming schemes (#823)

FreeNAS Ticket: #71675
Closes #596

Commit:
Fix renaming templates (#824)

This PR addresses the following:
- Change the release_snapshots structure from a list to a dict containing the snapshot name and the RELEASE dataset path it belongs too
- Strip unneeded whitespace on get_dataset_name
- Correctly rename the RELEASE snapshot of the target instead of trying a recursive, possible failure scenario.
- Set template readonly=off later down the line as other zfs set's were causing it to set back to readonly=on
- Loop through all the jails and update the source_template property if they have one

FreeNAS Ticket: #69942

Commit:
Migrate ip4 and ip6 from none to disable (#825)

- Add new method that will fix other properties we may need to address in the future

FreeNAS Ticket: #69276

Commit:
tkt-72067: Uppercase the RELEASE specified during create (#826)

* Uppercase the RELEASE specified during create

This helps avoid lazy sysadmin syndrome

Reported by: Michael W Lucas

FreeNAS Ticket: #72067

* Move this to lib

* Check for None

Commit:
Allow replacing an fstab with a different source and same destination (#827)

Also no longer throw an exception on the destination not existing, it will be created in the mount step.

FreeNAS Ticket: #72075
Closes #801

Commit:
Allow fixing broken fstabs with edit (#830)

- Also print the line number and problematic line in it's raw form.

FreeNAS Ticket: #72219
Closes #805

Commit:
Apply some source fstab validation when the fs type is nullfs (#829)

FreeNAS Ticket: #72203
Closes #806

Commit:
Bump to 1.1-RELEASE

Commit:
tkt-70993: Fix hostid_strict_check property value (#833)

* Fix hostid_strict_check property value

FreeNAS Ticket: #70993

* Drop a nested statement

Commit:
tkt-72619: Create a snapshot for update (#834)

* Create a snapshot for update

FreeNAS Ticket: #72619
Closes #618

* Use a more useful name for update snapshot

Commit:
Return correct devfs_ruleset for get -a (#835)

This also applies internally for any future usage

FreeNAS Ticket: #72641

Commit:
Use sys-rc style output for set (#836)

- Remove redundant logging from setting template=yes/no
- Return for setting template=no properly

FreeNAS Ticket: #72651
Closes #752

Commit:
tkt-72731: Allow regeneration of a new mac during clone (--newmac|-N) (#837)

* Allow regeneration of a new mac during clone (--newmac|-N)

- Cleanup snapshots after clone failure
- Correct a capital character

FreeNAS Ticket: #72731
Closes #783

* Handle cleanup for clones in lib

* Travis

* More unrelated flake8 changes

Commit:
tkt-72885: Force UTF-8 for stdout/stderr (#840)

* Force UTF-8 for stdout/stderr

FreeNAS Ticket: #72885
Closes #839

* Remove unused import

overhacked committed:
vnet_default_interface per bridge

Add ability to specify a third item in the
interfaces 'tuple', e.g. 'vnet0:bridge0:bge1'.

vnet_default_interface is still respected when
the third item is not present. An exception is
logged if the interface does not exist at parameter
set time or jail start time, just as
vnet_default_interfaces is checked.

Commit:
tkt-73236: Make booleans universal values (#842)

* Make booleans universal values

This greatly simplifies the user experience, in addition allows more boolean type properties to be easier added in the future with less regard to their parsing.

- All boolean properties can now be set as '1/0', 'on/off', 'yes/no', 'true/false'
- These now rest on disk as 1/0
- Fix some IP validation issues

FreeNAS Ticket: #73236

* Travis

Also ignore W504, that hurts readability

skarekrow reviewed (dismissed):
LGTM, thanks!

sonicaj committed:
Correct Locale settings

This commit forces locale settings to be 'en_US.UTF-8' for iocage fixing any unintended issues with user's locale settings.

skarekrow commented:
You will have to rebase this as well in addition to fixing the travis issues :+1:

sonicaj commented on iocage_lib/ioc_json.py:
```diff
@@ -2042,6 +2042,33 @@ def json_check_prop(self, key, value, conf):
+                                _callback=self.callback,
+                                silent=self.silent
+                            )
+                        if iface.count(':') == 2 and iface.split(':')[2] not in netifaces.interfaces():
```
@skarekrow I think this check should be done repeated on starting jail as well. It is possible that the interface no longer exists when a jail is started vs created. Thoughts ?

skarekrow replied:
@sonicaj Good point!

overhacked replied:
Agreed! I'll throw that in and rebase.

sonicaj replied:
Thank you ;)

sonicaj committed:
Merge pull request #845 from iocage/issues/ascii

Correct Locale settings

sonicaj reviewed (dismissed):
Thank you

skarekrow commented:
@overhacked All done with this one?

Commit:
Correct construct_truthy usage to use bitwise (#846)

FreeNAS Ticket: #73731

overhacked commented:
Reasonable question. I merged it into #843, because it made combining the netgraph and if_bridge code simpler. That said, it would only take a few minutes to remove the code for this feature from that PR so you can approve and merge separately, if you’d prefer?

Commit:
tkt-73740: Add ip_hostname support to iocage (#847)

* Add ip_hostname support to iocage

If a user sets their DNS records up and they match the jail name, this bool will allow the user to skip configuring the IP manually, as jail(8) will grab the IP from the resolver.

FreeNAS Ticket: #73740

* Allow fetching plugins and pkglist with this

- Users may still be able to footshoot as iocage does not verify they indeed have correctly set an IP address in their DNS.
- Also fix a bad error message

* Address Waqar's review

skarekrow commented:
@overhacked Preferrably we keep these separate, we'll merge this first so you can rebase the netgraph one on that. Makes the diff smaller. Btw a different PR was merged in the meantime, so netgraph's version for config will need to be 19.

overhacked commented:
Actually, I think #843 needs to be merged first, because this changes some of the interface creation functions that it refactors. It’ll be easier to rebase these changes onto the combined functions.

sonicaj committed:
RCTL for Iocage

This commit introduces following changes:
1) A class to manage RCTL related operations which are:
	a) Validate RCTL props
	b) Set RCTL rules
	c) Remove RCTL rules
	d) Check for existing RCTL rules
2) Ability to set RCTL props in the format 'action=amount'.
3) Setting RCTL rules after other start related operations are complete for a jail
4) Removing RCTL rules for a jail before we start any stopping operation for a jail when a jail stop is desired
5) The motivation behind 3 and 4 is that we should make sure start/stop behaviour is not affected by any RCTL rules at least for now as it can possibly affect the start/stop operations.
6) Allow setting/removing of RCTL rules for a jail which is running
7) There could be potentially more actions related to a single resource in RCTL rules, for now we only allow one. We can introduce support for multiple actions later gauging user reaction/usage.

Ticket: #73560

sonicaj committed:
Merge pull request #848 from iocage/issues/73560

tkt-73560: RCTL for Iocage

Commit:
Bump to 1.2-BETA

Commit:
tkt-74020: Remove unprotected devfs_rulesets (#849)

* Remove unprotected devfs_rulesets

This conditional was incorrect after recent changes to grabbing the dynamic ruleset

FreeNAS Ticket: #74020

* Use correct variable

* Remove unnecessary variables

* Missing f-string

Commit:
tkt-74253: Correct default bool props as well (#850)

* Correct default bool props as well

This will ensure that when we grab a single prop (like in ioc_start for dhcp) that they are actually 1/0 as expected.

FreeNAS Ticket: #74253

* Break out the write variable

This could have issues where write was True, but no properties needed to be fixed, setting it to False and never writing them.

Commit:
tkt-74281: Use hosts pkgs to remove pkgs for plugins (#851)

* Use hosts pkgs to remove pkgs for plugins

- Make revoked directory on plugin install
- Remove unused variable

FreeNAS Ticket: #74281

* Unrelated travis fixes

* Address Waqar's review

sonicaj committed:
Add Cpuset support to Iocage

This commit introduces the following changes:
1) A class to manage Cpuset prop wrt
  a) Validation
  b) Setting Cpuset Prop
2) Ability to set cpuset prop conforming to cpuset(1) format for '-l' flag only.
3) We only allow a very basic version of cpuset(1) support in which a user can specify a list of CPUs to apply to the jail.
4) When a jail is started, we set the cpuset prop in the system only at that point allowing the user to change the prop when the jail is running but not reflecting that change in the system. That change is only reflected in the system when jail is restarted. Motivation behind this step is that we are more likely to fail setting the prop when the jail is running in the system because of possible deadlock/validation issues raised by cpuset(1).
5) Point 3 and 4 can be revisited later to allow more support or enhance the functionality gauging user reaction/usage.

Ticket: #74235

sonicaj committed:
Fix locale for logging handlers

This commit fixes locale issues with logging handlers in CLI.
Closes #853
FreeNAS Ticket: #74623

sonicaj committed:
Merge pull request #857 from iocage/issues/74623

tkt-74623: Fix locale for logging handlers

sonicaj committed:
Merge pull request #852 from iocage/issues/74235

tkt-74235: Add Cpuset support to Iocage

overhacked committed:
Add debug option to CLI (#854)

Raises default logging level to VERBOSE (15) from DEBUG (10).
Adds command line option --debug/-D to re-enable debug logging.

Allows adding logging statements through code at level DEBUG or SPAM
that will not be displayed by default.

Commit:
tkt-74874: Show debug output of jail(8) during start/stop with the -D flag (#859)

* Show debug output of jail(8) during start/stop with the -D flag

This allows much easier debugging for users and developers.

FreeNAS Ticket: #74874

* Travis

olgeni committed:
Add missing markup to man page. (#860)

Commit:
tkt-72323: Drop dulwich dependency, add docurl support to ui.json for plugins (#862)

* Drop dulwich dependency, add docurl support to ui.json for plugins

- Replace Dulwich with GitPython
We were experiencing bad attribute usage with Dulwich, and while upstream has a commit to fix this, it didn't solve for our situation. Since we've had a couple problems with Dulwich in the past, let's try switching libraries to something that has been around longer.

- Also added docurl support to ui.json, this allows a plugin author to additionally specify where a user might be able to view documentation from the program authors themselves.

FreeNAS Ticket: #72323

* Be more consistent

* Remove last vestige

Commit:
IOCAGE_PLUGIN_IP now works with DHCP (#864)

DHCP grabbed it's IP too late as the jail was not running. Starting it now, and moving that IP logic.

FreeNAS Ticket: #75306

Commit:
Remove additional update snapshot from plugins (#865)

This came in a recent change to update, also needs to be nuked.

FreeNAS Ticket: #75333

Commit:
Create basejails as thickjails by default (#867)

- Remove a stale snapshot from thick jail creation
- Add flag -B to create that allows you to retain the original clonejail basejail creation behavior.

Closes #863
FreeNAS Ticket: #75360

Commit:
Exclude tests from dist (#866)

Closes #861
FreeNAS Ticket: #75351

Commit:
Add exec_created property (#868)

Description of the property:
exec.created
     Command(s) to run in the system environment right after a jail
     has been created, but before commands (or services) get executed
     in the jail.

Closes #855
FreeNAS Ticket: #75450

kpfleming committed:
Remove references to 'chroot' command from docs (#794)

* Remove references to 'chroot' command from docs

The 'chroot' command has been deprecated, so the docs should not refer to it
or suggest its use. For templates, this means that the only way to modify
a template is to convert it back into a jail, change it, and then convert to
a template again.

Closes issue #763

* Expand instructions for modifying templates.

skarekrow commented:
Sorry about that! Some git-fu backfired, can you please reopen this? Thanks!
